wasm/src/graphics/webgl.rs

```diff
@@ -0,0 +1,126 @@
+use super::webgl_bindings::Gl2;
+use super::GraphicsApi;
+use crate::error::ClientError;
+use rask_engine::{math::Mat3, resources::Texture};
+
+mod imports {
+    extern "C" {
+        pub fn get_canvas_size() -> u32;
+        pub fn set_canvas_size(w: u32, h: u32);
+    }
+}
+
+fn get_canvas_size() -> (u32, u32) {
+    let x = unsafe { imports::get_canvas_size() };
+    (x >> 16, x & 0xffff)
+}
+
+fn set_canvas_size(w: u32, h: u32) {
+    unsafe { imports::set_canvas_size(w, h) }
+}
+
+#[derive(Debug)]
+pub enum WebGl2Error {
+    ContextLost,
+    InvalidEnum,
+    InvalidValue,
+    InvalidOperation,
+    InvalidFramebufferOperation,
+    OutOfMemory,
+    UnknownError,
+}
+
+impl From<u32> for WebGl2Error {
+    fn from(v: u32) -> Self {
+        match v {
+            Gl2::INVALID_ENUM => WebGl2Error::InvalidEnum,
+            Gl2::INVALID_VALUE => WebGl2Error::InvalidValue,
+            Gl2::INVALID_OPERATION => WebGl2Error::InvalidOperation,
+            Gl2::INVALID_FRAMEBUFFER_OPERATION => WebGl2Error::InvalidFramebufferOperation,
+            Gl2::OUT_OF_MEMORY => WebGl2Error::OutOfMemory,
+            Gl2::CONTEXT_LOST_WEBGL => WebGl2Error::ContextLost,
+            _ => WebGl2Error::UnknownError,
+        }
+    }
+}
+
+impl std::fmt::Display for WebGl2Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                WebGl2Error::ContextLost => "lost webgl context",
+                WebGl2Error::InvalidEnum => "invalid enum",
+                WebGl2Error::InvalidValue => "invalid value",
+                WebGl2Error::InvalidOperation => "invalid operation",
+                WebGl2Error::InvalidFramebufferOperation => "invalid framebuffer operation",
+                WebGl2Error::OutOfMemory => "out of memory",
+                WebGl2Error::UnknownError => "unknown webgl2 error",
+            }
+        )
+    }
+}
```

Comment:
I feel like this belongs in some sort of error module. Either `error.rs` directly, or an new module. Anyho from implementation for client error should be derived

Reply:
I am going to split the webgl code into multiple modules in an own directory

Reply:
I'll try to add some unit test to the error conversions